components/AttributeAutocomplete/index.tsx

```diff
@@ -0,0 +1,96 @@
+import { SxProps } from '@mui/material'
+import { Autocomplete, Chip, TextField } from '@mui/material'
+import React from 'react'
+import { AttributeRequest, OptionsAttribute } from 'utils/types'
+
+interface AutocompleteAttributeOption {
+  id: string
+  label: string
+  value: string
+  color: string
+}
+
+interface Props {
+  attributes: OptionsAttribute[]
+  onChange?: (e: React.SyntheticEvent, attributes: AttributeRequest) => void
+  sx?: SxProps
+}
+
+function buildAutocompleteOptions(
+  attributes: OptionsAttribute[]
+): AutocompleteAttributeOption[] {
+  return attributes.reduce((acc, attribute) => {
+    const options = attribute.possibleValues.map((value) => ({
+      id: attribute._id!, // todo: remove when request and reponse objects are separated
+      label: attribute.name,
+      value,
+      color: attribute.color,
+    }))
+
+    return [...acc, ...options]
+  }, [] as AutocompleteAttributeOption[])
+}
+
+function buildAttributeRequest(
+  attributes: AutocompleteAttributeOption[]
+): AttributeRequest {
+  return attributes.reduce((acc, attribute) => {
+    const attrReq = {
+      [attribute.id]: attribute.value,
+    }
+
+    return { ...acc, ...attrReq }
+  }, {})
+}
+
+export default function AttributeAutocomplete({
+  attributes,
+  sx,
+  onChange,
+}: Props) {
+  const options = buildAutocompleteOptions(attributes)
+
+  return (
+    <Autocomplete
+      multiple
+      options={options}
+      groupBy={(option) => option.label}
+      getOptionLabel={(option) => option.value}
+      isOptionEqualToValue={(option, value) =>
+        option.value === value.value && option.label === value.label
+      }
+      renderTags={(tagValue, getTagProps) =>
+        tagValue.map((option, index) => (
+          <Chip
+            label={option.value}
+            style={{ backgroundColor: option.color }}
+            {...getTagProps({ index })}
+            key={option.id}
+          />
+        ))
+      }
+      renderInput={(params) => <TextField {...params} label="Attributes" />}
+      onChange={(e, attributes) => {
+        if (!attributes.length) {
+          if (!!onChange) onChange(e, {} as AttributeRequest)
+          return
+        }
+
+        // get and remove most recent attribute (the one that triggered this event)
+        const newAttr = attributes.pop() as AutocompleteAttributeOption
+
+        // check if there exists another attribute with the same label
+        const idx = attributes.findIndex(
+          (a) => a.label == newAttr.label && a.value != newAttr.value
+        )
+
+        // if so, replace that attribute with the new one. otherwise re-append it to the attributes array
+        if (idx != -1) attributes[idx] = newAttr
+        else attributes.push(newAttr)
+
+        if (!!onChange) onChange(e, buildAttributeRequest(attributes))
```

Comment:
Can you explain why `onChange` is optional?

Reply:
`onChange` is optional just in case there is a time when we just want to display the component without using any of it's underlying data. I got this idea from the MUI autocomplete component since it's `onChange` function is optional. Again, not sure this case will arise in our application, but it generalizes the functionality of the component without making any assumptions about its use.